### DISCUSSION TITLE
Write both vertex and cell centered fields in same cgns file

### DISCUSSION THREAD

**Discussion by eliesaikali** [CGNS/CGNS#745] [CGNStalk]
Hello,

I am wondering if it is possible to write cell-centered and vertex located fields in same file. I tried to used the existing test codes to try if it is possible but it is not working, I am only able to open the first ones (on vertices) but not the other. 
Any one can help ? here is the code ! Thanks

Best regards

--------------------------- 
```C
#include <iostream>
#include <stdio.h>
#include <string.h>
#include "cgnslib.h"

int main()
{
  int ni, nj, nk, i, j, k;
    {
      double x[9][17][21], y[9][17][21], z[9][17][21];
      cgsize_t isize[3][3];
      int index_file, icelldim, iphysdim, index_base;
      int index_zone, index_coord;
      char basename[33], zonename[33];

      ni = 21;
      nj = 17;
      nk = 9;
      for (k = 0; k < nk; ++k)
        for (j = 0; j < nj; ++j)
          for (i = 0; i < ni; ++i)
            {
              x[k][j][i] = i;
              y[k][j][i] = j;
              z[k][j][i] = k;
            }


      if (cg_open("grid_c.cgns", CG_MODE_WRITE, &index_file))
        cg_error_exit();

      strcpy(basename, "Base");
      icelldim = 3;
      iphysdim = 3;
      cg_base_write(index_file, basename, icelldim, iphysdim, &index_base);
      /* define zone name (user can give any name) */
      strcpy(zonename, "Zone  1");
      /* vertex size */
      isize[0][0] = 21;
      isize[0][1] = 17;
      isize[0][2] = 9;
      /* cell size */
      isize[1][0] = isize[0][0] - 1;
      isize[1][1] = isize[0][1] - 1;
      isize[1][2] = isize[0][2] - 1;
      /* boundary vertex size (always zero for structured grids) */
      isize[2][0] = 0;
      isize[2][1] = 0;
      isize[2][2] = 0;
      /* create zone */
      cg_zone_write(index_file, index_base, zonename, *isize, CGNS_ENUMV(Structured), &index_zone);
      /* write grid coordinates (user must use SIDS-standard names here) */
      cg_coord_write(index_file, index_base, index_zone, CGNS_ENUMV(RealDouble), "CoordinateX", x, &index_coord);
      cg_coord_write(index_file, index_base, index_zone, CGNS_ENUMV(RealDouble), "CoordinateY", y, &index_coord);
      cg_coord_write(index_file, index_base, index_zone, CGNS_ENUMV(RealDouble), "CoordinateZ", z, &index_coord);
      
      cg_close(index_file);
    }

    {
      double r[9][17][21], p[9][17][21];
      int index_file, index_base, index_zone, index_flow, index_field;
      char solname[33];

      ni = 21;
      nj = 17;
      nk = 9;
      for (k = 0; k < nk; k++)
        for (j = 0; j < nj; j++)
          for (i = 0; i < ni; i++)
            {
              r[k][j][i] = (float) i;
              p[k][j][i] = (float) j;
            }

      if (cg_open("grid_c.cgns", CG_MODE_MODIFY, &index_file))
        cg_error_exit();

      std::cerr << " index_file  : " << index_file << std::endl;

      index_base = 1;
      index_zone = 1;
      strcpy(solname, "FlowSolution_vert");
      cg_sol_write(index_file, index_base, index_zone, solname, CGNS_ENUMV(Vertex), &index_flow);
      cg_field_write(index_file, index_base, index_zone, index_flow, CGNS_ENUMV(RealDouble), "Density", r[0][0], &index_field);
      cg_field_write(index_file, index_base, index_zone, index_flow, CGNS_ENUMV(RealDouble), "Pressure", p[0][0], &index_field);

      cg_close(index_file);

    }

    {
      double r[8][16][20], p[8][16][20];
      int index_file, index_base, index_zone, index_flow, index_field;
      char solname[33];

      ni = 20;
      nj = 16;
      nk = 8;
      for (k = 0; k < nk; k++)
        for (j = 0; j < nj; j++)
          for (i = 0; i < ni; i++)
            {
              r[k][j][i] = (float) i;
              p[k][j][i] = (float) j;
            }

      if (cg_open("grid_c.cgns", CG_MODE_MODIFY, &index_file))
        cg_error_exit();
      index_base = 1;
      index_zone = 1;
      strcpy(solname, "FlowSolution_cent");
      cg_sol_write(index_file, index_base, index_zone, solname, CGNS_ENUMV(CellCenter), &index_flow);
      cg_field_write(index_file, index_base, index_zone, index_flow, CGNS_ENUMV(RealDouble), "Density_center", r[0][0], &index_field);
      cg_field_write(index_file, index_base, index_zone, index_flow, CGNS_ENUMV(RealDouble), "Pressure_center", p[0][0], &index_field);
      cg_close(index_file);
    }

  return 0;
}
```

**Comment by eliesaikali** [comment]
things are getting more complicated with time dependant flow solution ...

**Comment by MicK7** [comment]
Hello,

What do you mean by "I am only able to open the first ones (on vertices) but not the other." ? 
The code provided looks correct at first sight.

For time dependant flow solution, you can create one file per time step with its time information embedded or store all timestep in the same file. In this case you can create FlowSolution_t node with different names (for instance VertexSol_0005, VertexSol_0006 .... and  CelSol_0005, CellSol_0006 ... ) but you need to create Iterative FlowSolutionPointers in this case.

**Comment by eliesaikali** [comment]
Hello ! 
Thanks for your answer. In fact, the problem is that I am having issues when writing at the same time vertex/cell-centered fields into the same base/zone. 
I think that the problem is that I can't have two "cg_sol_write" applied at each time step for same base.

At present, i did the following trick. I duplicated the base so that I host vertex field on base 1 , zone 1 and center field on base 2, zone 1. Sure I have now 2 FlowSolution_t for each base, and it works now ... but I would appreciate if you tell be a better way to that without the need of duplicating the same base and the same zone twice ...

Regards and thanks again !
Elie Saikali

**Reply by MicK7** [reply]
If you have an existing base and zone in the file, it should be possible to have multiple cg_sol_write calls. Does it segfault or crash, in this case it might be a bug ?

When the solution are written, you need to create a VertexSolutionPointers and CellSolutionPointers which are FlowSolutionPointers node under the zone and store in each the list of name corresponding to solution at Vertex and Cell for each timestep. The list of timestep should be declared at the base level.

**Reply by eliesaikali** [reply]
Hello and thanks for your reply !
Its not working for me, I have already tried what you proposed but its not working. Can I have some help if i share my main.cpp ?
Thanks

**Reply by eliesaikali** [reply]
I wrote the fields in the following order : vertex then cell

By looking at the hdfview, i can see both fields and the different names of the solutions.
when i open in paraview, i see the base (zone, single one) both fields the cell and vertex one. 

when i choose the vertex field (the first one written), all Ok i can see it in paraview at all time steps written

when i select the cell field, nothing display in the field selections and nothing to visualize (but no crash)

Same in visit, when i select the vertex field all is OK, while it crashes completely when i select the cell field with an error : invalid field

Thanks

**Reply by jrwrigh** [reply]
Can you verify in `hdf5view` (or `cgnsview`) that the data is correctly written to the file? `cgnscheck` might also be useful.

**Reply by eliesaikali** [reply]
Every thing seems ok in the hdf5view but I really dont understand the issue ...

**Reply by jrwrigh** [reply]
The reason I ask for the other methods (`hdf5view` and `cgnscheck`) is that will verify that the *file* is correct. The file being correct and the file being fully readable by another application are unfortunately two different questions. 

CGNS file readers often have assumptions on the form of the CGNS file being read in, some being more flexible than others. For better or for worse, the CGNS standard is very flexible and broad and most reader implementations limit themselves on how exactly they read the files. The end result is having "valid" CGNS files to the CGNS spec while the reader implementations may not handle that specific file.

**Reply by jrwrigh** [reply]
> When the solution are written, you need to create a VertexSolutionPointers and CellSolutionPointers which are FlowSolutionPointers node under the zone and store in each the list of name corresponding to solution at Vertex and Cell for each timestep.

@MicK7 I'm assuming this goes off of this line from [this page of the spec](https://cgns.github.io/CGNS_docs_current/sids/timedep.html)?

> Any number of extra [DataArray_t](https://cgns.github.io/CGNS_docs_current/sids/data.html#DataArray) nodes are allowed. These should be used to record data not covered by this specification.

If that's the case, I doubt ParaView or VisIt supports that since the spec just has it as arbitrary data.

**Reply by jrwrigh** [reply]
I also assume you mean a structure like this:
`<CGNSBase_t>/<Zone_t>/<ZoneIterativeData_t>/CellSolutionPointers` and `<CGNSBase_t>/<Zone_t>/<ZoneIterativeData_t>/VertexSolutionPointers`?

Having a cell-centered and vertex centered field simultaneously is pretty common (often for pressure cell-centered and vertex for velocity) so I feel like this use case should probably be in the spec.

**Reply by jrwrigh** [reply]
@eliesaikali sent me a CGNS file. Here's the CGNSview of the file:
![image](https://github.com/CGNS/CGNS/assets/20801821/28be5469-ab8e-4059-a4a0-854595d42773)

There are a few minor issues that I can see (none of the `FlowSolutionVert` nodes don't have `GridLocation` associated with them, there's a `FlowSolutionPointers` node on `FlowSolutionVert1`).

After removing the misplaced `FlowSolutionPointers` node, `cgnscheck` only returns warnings about dataclasses, so it's compliant.

**Reply by eliesaikali** [reply]
I dont understand why there is no Gridlocation associated for the FlowSolutionVert, although I do : 

cg_sol_write(index_file, index_base, index_zone, sn1[n].c_str(), CGNS_ENUMV(Vertex), &index_flow_som)
and
cg_sol_write(index_file, index_base, index_zone, sn2[n].c_str(), CGNS_ENUMV(CellCenter), &index_flow_elem)

**Reply by MicK7** [reply]
It's ok if the FlowSolutionVert do not have a GridLocation because the CGNS standard assume that when there is none then it is Vertex .
It seems like you only have one solution pointer under ZoneIterativeData.
You can have a second one thus the first one can be use with FlowSolutionVert# while the last one will be for FlowSolutionCent#

**Reply by eliesaikali** [reply]
Hello !
I do have 2 solution pointers ! thats what I wrote

```C
  cg_array_write("FlowSolutionPointers", CGNS_ENUMV(Character), 2, idata, solname_vert.c_str());
  if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 1, "end"))
      cg_error_exit();

  cg_array_write("FlowSolutionPointers", CGNS_ENUMV(Character), 2, idata, solname_cent.c_str());
  if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 2, "end"))
      cg_error_exit();
```

**Reply by MicK7** [reply]
I would have expected something like that:

```C
if (cg_goto(index_file, index_base, "Zone_t", index_zone, "ZoneIterativeData", 1, "end"))
cg_error_exit();

cg_array_write("FlowSolutionPointers", CGNS_ENUMV(Character), 2, idata, solname_vert.c_str());
cg_array_write("FlowSolutionCellPointers", CGNS_ENUMV(Character), 2, idata, solname_cent.c_str());
```

**Reply by eliesaikali** [reply]
Yes ! thats the method where I do the iter stuff !

```C
void write_iters()
{
  int index_base = 1, index_zone = 1, nsteps = 3;
  double time[3];
  /* set up the times corresponding to the 3 solutions to be stored: */
  time[0] = 10.;
  time[1] = 20.;
  time[2] = 50.;

  cg_biter_write(index_file, index_base, "TimeIterValues", nsteps);

  cg_goto(index_file, index_base, "BaseIterativeData_t", 1, "end");
  cgsize_t idata[2], nuse = 3;
  cg_array_write("TimeValues", CGNS_ENUMV(RealDouble), 1, &nuse, &time);

  cg_ziter_write(index_file, index_base, index_zone, "ZoneIterativeData");

  cg_goto(index_file, index_base, "Zone_t", index_zone, "ZoneIterativeData_t", 1, "end");

  idata[0] = 32;
  idata[1] = 3;
  cg_array_write("FlowSolutionPointers", CGNS_ENUMV(Character), 2, idata, solname_vert.c_str());
  if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 1, "end"))
      cg_error_exit();

  cg_array_write("FlowSolutionPointers", CGNS_ENUMV(Character), 2, idata, solname_cent.c_str());
  if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 2, "end"))
      cg_error_exit();

  cg_simulation_type_write(index_file, index_base, CGNS_ENUMV(NonTimeAccurate));
}
```

**Reply by MicK7** [reply]
You should not have
```C
if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 1, "end"))
cg_error_exit();
```

**Reply by eliesaikali** [reply]
It changed nothing ... same issue ...

**Reply by MicK7** [reply]
With the following code you should see two ***Pointers under ZoneIterativeData

````C
void write_iters()
{
  int index_base = 1, index_zone = 1, nsteps = 3;
  double time[3];
  /* set up the times corresponding to the 3 solutions to be stored: */
  time[0] = 10.;
  time[1] = 20.;
  time[2] = 50.;

  cg_biter_write(index_file, index_base, "TimeIterValues", nsteps);

  cg_goto(index_file, index_base, "BaseIterativeData_t", 1, "end");
  cgsize_t idata[2], nuse = 3;
  cg_array_write("TimeValues", CGNS_ENUMV(RealDouble), 1, &nuse, &time);

  cg_ziter_write(index_file, index_base, index_zone, "ZoneIterativeData");

  cg_goto(index_file, index_base, "Zone_t", index_zone, "ZoneIterativeData_t", 1, "end");

  idata[0] = 32;
  idata[1] = 3;
  cg_array_write("FlowSolutionPointers", CGNS_ENUMV(Character), 2, idata, solname_vert.c_str());

  cg_array_write("FlowSolutionCellPointers", CGNS_ENUMV(Character), 2, idata, solname_cent.c_str());

  cg_simulation_type_write(index_file, index_base, CGNS_ENUMV(NonTimeAccurate));

  if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 1, "end"))
      cg_error_exit();

  if (cg_goto(index_file, index_base, "Zone_t", index_zone, "FlowSolution_t", 2, "end"))
      cg_error_exit();
}
````

**Reply by MicK7** [reply]
If you are using ParaView to display results you can also activate the small wheel and check "UseUnsteadyPattern". This allows to scan the FlowSolution names and get the iteration number from the end of the name.

**Reply by eliesaikali** [reply]
<img width="478" alt="image" src="https://github.com/CGNS/CGNS/assets/38346555/70cd3562-ee74-4e8c-8802-f819a21c4d19">


Here is the screenshot, as before, the center pressure fields appears in the file but not possible to select and visualize it

**Reply by eliesaikali** [reply]
only the density (vertex one) is available

**Reply by eliesaikali** [reply]
So I can assume that this is a bug in CGNS and that it is not possible to write both locations to same Zone in same file ?

**Reply by MicK7** [reply]
Can you share a small file generated with this procedure to check if it compliant with the CGNS standard ?
It can be either a bug in CGNS lib or a bug in ParaView.

**Reply by eliesaikali** [reply]
Thanks Mickael for your reply !

Sure, here is both the cgns and the main.cpp file used to generate the file (changed extension to txt so that I can upload them here ...)
Thanks for your help

[test_vertex_cellcenter.cgns.txt](https://github.com/CGNS/CGNS/files/13906868/test_vertex_cellcenter.cgns.txt)
[main8.cpp.txt](https://github.com/CGNS/CGNS/files/13906869/main8.cpp.txt)

**Reply by MicK7** [reply]
Ok I found the issue. To make it work you need to name the FlowSolutionPointers with the pattern FlowSolutionXXXPointers.
In your case I modified the file to rename CellSolutionPointers to FlowSolutionCellPointers and it works perfectly now.

**Reply by MicK7** [reply]
I also tried in ParaView with the small wheel to Ignore FlowSolutionPointers and activate Unsteady Pattern on Flowsolution_t and it also did the job.

**Reply by eliesaikali** [reply]
Ah super thanks a lot for your help.
I will try when i go back to my office. So you just changed in the method write_iters() in the cpp code ? Or in the write solutions too ?

**Reply by eliesaikali** [reply]
Hello, I tested in the C++ code to do what you proposed. Now in paraview i can see both fields center and vertex. But the fields values dont change by changing the time (unless I activate the Unsteady Pattern on Flowsolution_t). But I would appreciate if I can do it without activating this in paraview.

Besides, it crashes always with VisIt ...

**Reply by eliesaikali** [reply]
Mickael, can you please share the corrected main.cpp please ?

**Reply by MicK7** [reply]
[main_working.cpp.txt](https://github.com/CGNS/CGNS/files/13939339/main_working.cpp.txt)

**Reply by eliesaikali** [reply]
Oh thank you so much Mickael, I confirm that it works in Paraview now ... but continues crashing in VisIt ....
maybe the CGNS plugin in Visit needs to be modified ?

Really thanks because its not trivial to know that we should keep **FlowSolutionPointers** for first solution and **FlowSolutionCellPointers** for second. I thought it should be **FlowSolutionVertPointers** and **FlowSolutionCellPointers** ...

**Comment by MicK7** [comment]
> I also assume you mean a structure like this: `<CGNSBase_t>/<Zone_t>/<ZoneIterativeData_t>/CellSolutionPointers` and `<CGNSBase_t>/<Zone_t>/<ZoneIterativeData_t>/VertexSolutionPointers`?
> 
> Having a cell-centered and vertex centered field simultaneously is pretty common (often for pressure cell-centered and vertex for velocity) so I feel like this use case should probably be in the spec.

@jrwrigh  Yes exactly what I meant.
It is already supported by ParaView with a restriction to FlowSolutionXXXPointers naming.
So the current support is:
`<CGNSBase_t>/<Zone_t>/<ZoneIterativeData_t>/FlowSolutionPointers` 
`<CGNSBase_t>/<Zone_t>/<ZoneIterativeData_t>/FlowSolutionCellPointers`
